[SelectionDAG] Handle more opcodes in isGuaranteedNotToBeUndefOrPoison

bjope opened the pull request:
Add special handling of EXTRACT_SUBVECTOR, INSERT_SUBVECTOR, EXTRACT_VECTOR_ELT, INSERT_VECTOR_ELT and SCALAR_TO_VECTOR in isGuaranteedNotToBeUndefOrPoison. Make use of DemandedElts to improve the analysis and only check relevant elements for each operand.

Also start using DemandedElts in the recursive calls that check isGuaranteedNotToBeUndefOrPoison for all operands for operations that do not create undef/poison. We can do that for a number of elementwise operations for which the DemandedElts can be applied to every operand (e.g. ADD, OR, BITREVERSE, TRUNCATE).

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5456,6 +5456,85 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
+                                                             PoisonOnly,
+                                                             Depth + 1))
+      return false;
+    return true;
```
can all of this be simplified to the following?
```cpp
return DemandedElts == 1 && isGuaranteedNotToBeUndefOrPoison(Op.getOperand(0), PoisonOnly, Depth + 1);
```

bjope replied:
Consider PoisonOnly==true and only the upper undef elements are demanded. We should be allowed to return true indicating that the upper elements aren't poison. Although not sure if that would be a very rare situation.

There is also the odd case when DemandedElts==0 for which we now return true. I'm basically same logic as for operations like INSERT_SUBVECTOR and VECTOR_SHUFFLE here. That we first check !isGuaranteedNot for values that map to the demanded elements. If we passed all those checks then isGuaranteedNot should be true for all demanded elements and we return true. And in case DemandedElts==0 we also end up returning true.

RKSimon reviewed (commented):
Is the peekThroughInsertVectorElt code necessary for the test changes? I'm not opposed to adding handling for many opcodes without much test coverage but the INSERT_VECTOR_ELT code looks to be a very special case.

RKSimon reviewed (commented):
Can you seperate off the peekThroughInsertVectorElt change into a followup PR? I think that's the only thing causing hesitation on this patch - the subvector handling in particular looks trivial.

arsenm reviewed (commented):
Can you fix the title to say how it's improved? This is just adding handling of new ops

RKSimon reviewed (commented):
Cheers - lack of test coverage is annoying, but par for the course with DAG freeze handling.

I just need to get #149797 done first as I'm not going to be able to create a proper fix for #148084 that I can merge into 21.x.

RKSimon commented:
@bjope please can you merge against trunk?

bjope committed:
[SelectionDAG] Handle more opcodes in isGuaranteedNotToBeUndefOrPoison

Add special handling of EXTRACT_SUBVECTOR, INSERT_SUBVECTOR,
EXTRACT_VECTOR_ELT, INSERT_VECTOR_ELT and SCALAR_TO_VECTOR in
isGuaranteedNotToBeUndefOrPoison. Make use of DemandedElts to
improve the analysis and only check relevant elements for each
operand.

Also start using DemandedElts in the recursive calls that
check isGuaranteedNotToBeUndefOrPoison for all operands for
operations that do not create undef/poison. We can do that for
a number of elementwise operations for which the DemandedElts
can be applied to every operand (e.g. ADD, OR, BITREVERSE, TRUNCATE).

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5458,6 +5458,83 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
+    EVT VT = InVec.getValueType();
+    auto *IndexC = dyn_cast<ConstantSDNode>(EltNo);
+    if (IndexC && VT.isFixedLengthVector() &&
+        IndexC->getZExtValue() < VT.getVectorNumElements()) {
```
`IndexC->getAPIntValue().ult(VT.getVectorNumElements())` (for consistency)

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5480,6 +5557,52 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
+  case ISD::UMAX:
+  case ISD::ZERO_EXTEND:
+  case ISD::SIGN_EXTEND:
+  case ISD::ANY_EXTEND:
```
Should ANY_EXTEND be here (canCreateUndefOrPoison will always return true)?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5480,6 +5557,52 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
+           isGuaranteedNotToBeUndefOrPoison(Op.getOperand(0), DemandedElts,
+                                            PoisonOnly, Depth + 1);
+
+  case ISD::BSWAP:
```
Which of these have you explicitly checked and which are just by inspection?

bjope replied:
(I've been out of office some days so memory is a bit vague...)

I think I added most of these just by trying to find fairly common opcodes (at least pretty common from my perspective) that have "elementwise operands".

Unfortunately this patch does not include any lit tests that is impacted by this extra DemandedElts propagation. The motivating case was something I think I saw when working on the other patches with "DoNotPoisionEltMask" for SimplifyDemanded functions.

bjope commented:
Ping. Ok to merge this now?

RKSimon reviewed (approved):
LGTM with one minor

RKSimon committed:
Merge branch 'main' into users/bjope/prepare_2
